bundler/bundler.go

```diff
@@ -50,49 +73,144 @@ func BundleBytes(bytes []byte, configuration *datamodel.DocumentConfiguration) (
-			mappedReference := mappedReferences[sequenced.FullDefinition]
-
-			// if we're in the root document, don't bundle anything.
-			refExp := strings.Split(sequenced.FullDefinition, "#/")
-			if len(refExp) == 2 {
-				if refExp[0] == sequenced.Index.GetSpecAbsolutePath() || refExp[0] == "" {
-					if root && !inline {
-						idx.GetLogger().Debug("[bundler] skipping local root reference",
-							"ref", sequenced.Definition)
-						continue
-					}
-				}
+	idx := rolodex.GetRootIndex()
+	mappedReferences := idx.GetMappedReferences()
+	sequencedReferences := idx.GetRawReferencesSequenced()
+
+	for _, sequenced := range sequencedReferences {
+		mappedReference := mappedReferences[sequenced.FullDefinition]
+		if mappedReference == nil {
+			return nil, fmt.Errorf("no mapped reference found for: %s", sequenced.FullDefinition)
+		}
+
+		if mappedReference.DefinitionFile() == idx.GetSpecAbsolutePath() {
+			// Don't bundle anything that's in the main file.
+			continue
+		}
+
+		switch opts.RelativeRefHandling {
+		case RefHandlingInline:
+			// Just deal with simple inlining.
+			sequenced.Node.Content = mappedReference.Node.Content
+		case RefHandlingCompose:
+			// Recursively collect all reference targets to be bundled into the root
+			// file.
+			bundledComponents := make(map[string]*index.ReferenceNode)
+			if err := bundleRefTarget(sequenced, mappedReference, bundledComponents, opts); err != nil {
+				return nil, err
 			}
 
-			if mappedReference != nil && !mappedReference.Circular {
-				sequenced.Node.Content = mappedReference.Node.Content
-				continue
+			model, err = composeDocument(model, bundledComponents)
+			if err != nil {
+				return nil, err
 			}
+		}
+	}
+
+	return model.Render()
+}
+
+func composeDocument(model *highV3.Document, comps map[string]*index.ReferenceNode) (*highV3.Document, error) {
+	lowModel := model.GoLow()
+
+	components := lowModel.Components
+
+	for def, component := range comps {
+		defParts := strings.Split(def, "/")
+		// TODO: use constant from low model labels
+		if len(defParts) != 4 || defParts[1] != lowV3.ComponentsLabel {
+			return nil, fmt.Errorf("unsupported component section: %s", def)
+		}
+		spew.Dump(component)
 
-			if mappedReference != nil && mappedReference.Circular {
-				if idx.GetLogger() != nil {
-					idx.GetLogger().Warn("[bundler] skipping circular reference",
-						"ref", sequenced.FullDefinition)
-				}
+		switch defParts[2] {
+		case "schemas":
+			key := low.KeyReference[string]{
+				Value: defParts[3],
+				KeyNode: &yaml.Node{
+					Kind:  yaml.ScalarNode,
+					Style: yaml.TaggedStyle,
+					Tag:   "!!str",
+					Value: defParts[3],
+				},
+			}
+			value := low.ValueReference[*base.SchemaProxy]{
+				Reference: low.Reference{},
+				Value: &base.SchemaProxy{
+					Reference: low.Reference{},
+					NodeMap:   &low.NodeMap{Nodes: &sync.Map{}},
+				},
+				ValueNode: &yaml.Node{},
 			}
+			components.Value.Schemas.Value.Set(key, value)
```

Comment:
It would appear, I need to put together the `key` and `value` so I can set them in the ordered map for Schemas. However, with all the `Value`, `ValueNode`, `Reference` etc. fields, I really don't understand what I'd be doing here. 😅 Can you once again point me in the right direction, please?